jedi_language_server/server.py

```diff
@@ -724,6 +738,149 @@ def did_change_configuration(
     """
 
 
+def _raw_semantic_token(
+    server: JediLanguageServer, n: Name
+) -> Union[list[int], None]:
+    """Find an appropriate semantic token for the name.
+
+    This works by looking up the definition (using jedi ``goto``) of the name and
+    matching the definition's type to one of the availabile semantic tokens. Further
+    improvements are possible by inspecting context, e.g. semantic token modifiers such
+    as ``abstract`` or ``async`` or even different tokens, e.g. ``property`` or
+    ``method``. Dunder methods may warrant special treatment/modifiers as well.
+    The return is a "raw" semantic token rather than a "diff." This is in the form of a
+    length 5 array of integers where the elements are the line number, starting
+    character, length, token index, and modifiers (as an integer whose binary
+    representation has bits set at the indices of all applicable modifiers).
+    """
+    definitions: list[Name] = n.goto(
+        follow_imports=True,
+        follow_builtin_imports=True,
+        only_stubs=False,
+        prefer_stubs=False,
+    )
+    if not definitions:
+        server.show_message_log(
+            f"no definitions found for name \"{n.description}\" of type '{n.type}' ({n.line}:{n.column})",
+            MessageType.Debug,
+        )
+        return None
+
+    if len(definitions) > 1:
+        msg = (
+            f"multiple definitions found for name \"{n.description}\" of type '{n.type}' ({n.line}:{n.column}):\n"
+            f" {'\n'.join(map(lambda n: str(n), definitions))}"
+        )
+        server.show_message_log(msg, MessageType.Debug)
+
+    definition, *_ = definitions
+    if (
+        definition_type := SEMANTIC_TO_TOKEN_ID.get(definition.type, None)
+    ) is None:
+        server.show_message_log(
+            f"no matching semantic token for \"{n.description}\" of type '{n.type}' ({n.line}:{n.column})",
+            MessageType.Debug,
+        )
+        return None
+
+    return [n.line - 1, n.column, len(n.name), definition_type, 0]
+
+
+@SERVER.thread()
+@SERVER.feature(
+    TEXT_DOCUMENT_SEMANTIC_TOKENS_FULL,
+    SemanticTokensLegend(
+        token_types=SUPPORTED_SEMANTIC_TYPES, token_modifiers=[]
```

Comment:
As I mentioned earlier, `SUPPORTED_SEMANTIC_TYPES` should probably just be the values in `SEMANTIC_TO_TOKEN_ID`, since other values will never be present in the output of `_raw_semantic_token`